chainer/optimizer.py

```diff
@@ -652,9 +655,58 @@ def serialize(self, serializer):
             if rule is not None:
                 rule.serialize(serializer[name])
 
+    def loss_scaling(self, interval=1000, scale=None):
+        """Sets loss scaling configurations.
+
+        Args:
+            interval (integer): Number of iterations until scaling factor gets
+                doubled. This is effective when "dynamic" loss scaling is used.
+            scale (float): Loss scaling factor. If None, "dynamic" loss scaling
+                is used, otherwise "static" loss scaling is used.
+        """
+        if scale is None:
+            self._loss_scaling_mode = 'dynamic'
+            if interval < 1:
+                raise ValueError("interval must be greater equal to 1")
+            self._loss_scale = 1.0
+            self._loss_scaling_multiplier = math.pow(2.0, 1.0 / interval)
+            self._loss_scaling_isnan_ever = False
+        else:
+            self._loss_scaling_mode = 'static'
+            if scale <= 0:
+                raise ValueError("loss_scale must be positive number")
+            self._loss_scale = scale
+
     def set_loss_scale(self, loss_scale):
         """Sets loss scaling factor."""
-        self._loss_scale = loss_scale
+        self.loss_scaling('static', loss_scale)
+
+    def is_safe_to_update(self):
+        if self._loss_scaling_mode is not 'dynamic':
+            return True
+        is_safe = True
+        for name, param in self.target.namedparams():
+            xp = param.device.xp
+            if not xp.all(xp.isfinite(param.grad)):
+                is_safe = False
+                self._loss_scaling_isnan_ever = True
+                warnings.warn(
+                    'Non finite number found in param.grad of {}'
+                    ' (t:{}, loss_scale:{})'
+                    ''.format(name, self.t, self._loss_scale))
```

Comment:
Should you not break from the loop after this line in favor of performance?

Reply:
I did not put "break" here intentionally, since I thought it was better to show all layers which have 'NaN' in its grads.

Reply:
I see, could you elaborate a bit in what sense that'd be useful to the user? I though the performance impact was more relevant here.

Reply:
Actually "break" helps reduce the amount of computation, but it is only when "NaN" is detected in the gradients and it is not a common case.  For example, if the interval is 1000, the probability of detecting NaN is about 0.1%.  Thus, I think it is not so important whether to break the loop or not in terms of performance.

Reply:
That makes sense, thanks for the explanation. Let's leave it as is.